platforms/app_link_platform_iphone.js

```diff
@@ -16,19 +16,41 @@
  *    A fallback URL, if we can't direct the user somewhere better.
  */
 /*global app_link_is_path_whitelisted*/
-function app_link_platform_iphone(platform, fallback_url) {
+function app_link_platform_iphone(platform, fallback_url, referrer) {
   fallback_url = platform.store_url || fallback_url;
   var app_url = getAppUrl();
+
+  /*
+   * Attaches the referrer to the destination URL.
+   */
+  function attachReferrer(url, referrer) {
+    var destURL = url;
+    if (destURL.indexOf('referrer') == -1 && referrer.length > 0) {
+      if (destURL.indexOf('?') > -1) {
+        destURL = destURL + '&referrer=' + referrer;
+      }
+      else {
+        destURL = destURL + '?referrer=' + referrer;
+      }
+      return destURL;
+    }
+  }
+
+  // Attach 'referrer' to URLs if necessary.
+  if (referrer) {
+    fallback_url = attachReferrer(fallback_url, referrer);
+    app_url = attachReferrer(app_url, referrer);
```

Comment:
Not every App URL will want the referrer appended, in fact in may break some.  Let's put this logic behind the `platform.supports_qs`.  Same goes for other platform JS.